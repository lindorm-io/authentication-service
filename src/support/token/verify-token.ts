import { Client } from "@lindorm-io/koa-client";
import { Device } from "../../entity";
import { IAuthContext } from "../../typing";
import { ITokenIssuerVerifyData } from "@lindorm-io/jwt";
import { JWT_ISSUER } from "../../config";

export interface IGetDecodedTokenOptions {
  audience: string;
  client?: Client;
  device?: Device;
  token: string;
}

export const verifyToken = (ctx: IAuthContext) => (options: IGetDecodedTokenOptions): ITokenIssuerVerifyData => {
  const { logger, issuer } = ctx;
  const { tokenIssuer } = issuer;
  const { audience, client, device, token } = options;

  logger.debug("decoding token", {
    audience,
    client: client?.id,
    device: device?.id,
    token,
  });

  return tokenIssuer.verify({
    audience,
    clientId: client?.id,
    deviceId: device?.id,
    issuer: JWT_ISSUER,
    token,
  });
};
