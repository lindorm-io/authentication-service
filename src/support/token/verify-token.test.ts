import MockDate from "mockdate";
import { Client } from "@lindorm-io/koa-client";
import { ITokenIssuerSignData } from "@lindorm-io/jwt";
import { getTestClient, getTestIssuer, logger } from "../../test";
import { verifyToken } from "./verify-token";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("verifyToken", () => {
  let ctx: any;
  let client: Client;
  let signData: ITokenIssuerSignData;

  beforeEach(() => {
    const issuer = getTestIssuer();

    ctx = {
      logger,
      issuer: { tokenIssuer: issuer },
      metadata: { deviceId: "deviceId" },
    };

    client = getTestClient();

    signData = issuer.sign({
      audience: "audience",
      expiry: "5 minutes",
      subject: "subject",
      clientId: client.id,
    });
  });

  test("should verify token", () => {
    expect(
      verifyToken(ctx)({
        audience: "audience",
        client,
        token: signData.token,
      }),
    ).toMatchSnapshot();
  });

  test("should verify token without client", () => {
    expect(
      verifyToken(ctx)({
        audience: "audience",
        token: signData.token,
      }),
    ).toMatchSnapshot();
  });
});
