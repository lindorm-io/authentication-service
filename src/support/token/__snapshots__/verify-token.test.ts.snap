// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verifyToken should verify token 1`] = `
Object {
  "authContextClass": null,
  "authMethodsReference": null,
  "clientId": "9ba5e0c5-f2aa-4706-a27d-66369b547b0b",
  "deviceId": null,
  "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "level": 0,
  "payload": Object {},
  "permission": null,
  "scope": null,
  "subject": "subject",
}
`;

exports[`verifyToken should verify token without client 1`] = `
Object {
  "authContextClass": null,
  "authMethodsReference": null,
  "clientId": "9ba5e0c5-f2aa-4706-a27d-66369b547b0b",
  "deviceId": null,
  "id": "be3a62d1-24a0-401c-96dd-3aff95356811",
  "level": 0,
  "payload": Object {},
  "permission": null,
  "scope": null,
  "subject": "subject",
}
`;
